### PR TITLE
[#477] feat: support stage recompute for spark clients

### LINKED ISSUE #477: [FEATURE] Support stage recompute for Spark clients
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

Once `Uniffle`'s shuffle data read fails, the spark client has its chance to recompute the whole stage.

### Motivation

In a distributed cluster with large number of shuffler servers, it's common for some node to go down, such as:
1. node crash/maintenance due to hardware failures or security patches
2. Pod eviction if deployed in a k8s environment
3. vm/spot instance eviction if deployed in a cloud environment

Uniffle already has provide a mechanism to overcome this issue: the quorum protocol. But it requires multiple replica of the same shuffle data, which increases the network traffic and memory pressure on shuffle server.  And the E2E performance may be degraded due to the replication.

I'd like to provide a new way to alleviate the potential node failure(in rare chance). Once the whole stage could be recompute, the Spark App could be resilient to shuffle server node failure.

### Describe the solution

TBD.

A design doc would be added later

### Additional context

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by advancedxy**
### What changes were proposed in this pull request?
1. add shuffle manager service definition
2. shuffle manager server and client impl
3. draft spark stage recompute support

### Why are the changes needed?
To improve stability of Spark application over RSS
Resolve #477 

### Does this PR introduce _any_ user-facing change?
Yes. Updated lated

### How was this patch tested?
WIP

### CODE REVIEW

**Comment by jerqi** [comment]
@smallzhongfeng Could you help me review this pr?

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleManagerGrpcClient.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  public String getDesc() {
+    return "Shuffle manager grpc client ref " + host + ":" + port;
+  }
+  @Override
+  public RssReportShuffleFetchFailureResponse reportShuffleFetchFailure(RssReportShuffleFetchFailureRequest request) {
+    RssProtos.ReportShuffleFetchFailureRequest protoRequest = request.toProto();
+    int retryNum = 0;
+    while (retryNum <= maxRetryAttempts) {
```
Comment (smallzhongfeng) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleManagerGrpcClient.java:63]: What about using `RetryUtils` ?

**Review by smallzhongfeng — COMMENTED** [review]
Thanks for pinging me. But I need some time to read the design documents, and I will review them as soon as possible.

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleManagerGrpcClient.java (hunk 1 above)
Comment (smallzhongfeng) [reply internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleManagerGrpcClient.java:63]: I think the overall design idea is OK. I think you can continue.  But I still want to ask @xianjingfeng‘s opinion.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssFeatureMatrix.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle;
+
+public class RssFeatureMatrix {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssFeatureMatrix.java:20]: Could we move this method to `RssSparkShuffelUtils`?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkConfig.java [hunk 3]
```diff
@@ -265,6 +266,17 @@ public class RssSparkConfig {
                    + " spark.rss.estimate.server.assignment.enabled"))
       .createWithDefault(RssClientConfig.RSS_ESTIMATE_TASK_CONCURRENCY_PER_SERVER_DEFAULT_VALUE);
 
+  public static final ConfigEntry<Integer> RSS_SHUFFLE_MANAGER_GRPC_PORT = createIntegerBuilder(
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkConfig.java:269]: We need to add document for this feature.

File: proto/src/main/proto/Rss.proto [hunk 4]
```diff
@@ -488,4 +488,26 @@ message CancelDecommissionRequest {
...
+
+message ReportShuffleFetchFailureRequest {
+  // appId normally should be omitted, it's used to avoid wrongly request issued from remaining executors of another
+  // app which accidentally has the same shuffle manager port with this app.
+  string appId = 1;
+  int32 shuffleId = 2;
+  int32 stageAttemptId = 3;
+  int32 partitionId = 4;
+  string exception = 5;
+  string serverId = 6;
```
Comment (jerqi) [inline proto/src/main/proto/Rss.proto:506]: What do the serverId mean?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/reader/RssShuffleReader.java [hunk 5]
```diff
@@ -173,6 +175,20 @@ public Void apply(TaskContext context) {
     if (!(resultIter instanceof InterruptibleIterator)) {
       resultIter = new InterruptibleIterator<>(context, resultIter);
     }
+    // stage re-compute and shuffle manager server port are both set
+    if (rssConf.getBoolean(RssClientConfig.RSS_STAGE_RECOMPUTE, false)
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/reader/RssShuffleReader.java:179]: Is it right? Should we use Spark's config option?

File: client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java [hunk 6]
```diff
@@ -86,4 +86,6 @@ public class RssClientConfig {
   public static final String RSS_ESTIMATE_TASK_CONCURRENCY_PER_SERVER = "rss.estimate.task.concurrency.per.server";
   public static final int RSS_ESTIMATE_TASK_CONCURRENCY_PER_SERVER_DEFAULT_VALUE = 80;
 
+  public static final String RSS_STAGE_RECOMPUTE = "rss.stage.reCompute";
```
Comment (jerqi) [inline client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java:89]: `reCompute` seems weird.

File: internal-client/src/main/java/org/apache/uniffle/client/factory/ShuffleManagerClientFactory.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  private ClientType clientType;
+
+  public ShuffleManagerClientFactory(ClientType clientType) {
+    this.clientType = clientType;
+  }
+
+  public ShuffleManagerGrpcClient createShuffleManagerClient(String host, int port) {
+    if (clientType.equals(ClientType.GRPC)) {
```
Comment (jerqi) [inline internal-client/src/main/java/org/apache/uniffle/client/factory/ShuffleManagerClientFactory.java:38]: Will it have problems when we use `ClientType.GRPC_NETTY`?

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/AbstractRssShuffleManagerBase.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.spark.MapOutputTracker;
+import org.apache.spark.MapOutputTrackerMaster;
+import org.apache.spark.SparkEnv;
+import org.apache.spark.SparkException;
+import org.apache.spark.shuffle.RssFeatureMatrix;
+import org.apache.spark.shuffle.SparkVersionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractRssShuffleManagerBase implements RssShuffleManagerBase {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/AbstractRssShuffleManagerBase.java:33]: Why do we choose to add interface?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkConfig.java [hunk 9]
```diff
@@ -265,6 +266,17 @@ public class RssSparkConfig {
...
+      new ConfigBuilder(SPARK_RSS_CONFIG_PREFIX + RssClientConf.SHUFFLE_MANAGER_GRPC_PORT.key())
+          .internal()
+          .doc(RssClientConf.SHUFFLE_MANAGER_GRPC_PORT.description())
+  ).createWithDefault(-1);
+
+  public static final ConfigEntry<Boolean> RSS_STAGE_RECOMPUTE = createBooleanBuilder(
```
Comment (zuston) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkConfig.java:275]: Could you help put the new client config into `RssClientConf`?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java [hunk 10]
```diff
@@ -101,7 +101,8 @@ public boolean hasNext() {
       // read next segment
       long startFetch = System.currentTimeMillis();
       // depends on spark.shuffle.compress, shuffled block may not be compressed
-      CompressedShuffleBlock rawBlock = shuffleReadClient.readShuffleBlockData();
+      CompressedShuffleBlock rawBlock = null;
```
Comment (zuston) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:104]: Why?

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/AbstractRssShuffleManagerBase.java (hunk 8 above)
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/AbstractRssShuffleManagerBase.java:33]: As a shareable class for spark2.x and spark3.x ?

Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/AbstractRssShuffleManagerBase.java:33]: Em .... could we give a better name for this class?

File: proto/src/main/proto/Rss.proto (hunk 4 above)
Comment (jerqi) [reply proto/src/main/proto/Rss.proto:506]: We already have a proto type called ShuffleServerId. Should we reuse it?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssFeatureMatrix.java (hunk 2 above)
Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssFeatureMatrix.java:20]: I'm a little confused about this class name. FeatureMatrix sounds like a concept of linear algebra.

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/AbstractRssShuffleManagerBase.java (hunk 8 above)
Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/AbstractRssShuffleManagerBase.java:34]: How about
 `RssShuffleManagerBase` -> `RssShuffleManagerInterface`?
`AbstractRssShuffleManagerBase` -> `AbstractRssShuffleManager`?

File: proto/src/main/proto/Rss.proto (hunk 4 above)
Comment (jerqi) [reply proto/src/main/proto/Rss.proto:506]: It's difficult for us to change old protocol field. If it won't bring too much cost, I prefer that we have a unify style.

**Pull request closed without merging**